dbt/include/sqlite/macros/utils/datediff.sql

```diff
@@ -1,42 +1,89 @@
+-- These adjustments are particularly crucial for broader date parts like 'week'. 
+-- The decision to use CEIL or FLOOR for rounding was refined through testing to accurately reflect differences when dates are separated by as little as one millisecond.
+-- This empirical approach ensures that the macro not only meets expected functional accuracy but also addresses real-world use cases effectively. 
+-- For example, calculating the week difference between two timestamps only a millisecond apart requires careful consideration of how to round fractional weeks. 
+-- The choice to use CEIL in certain contexts was driven by the need to acknowledge even the minimal time difference as a full unit where contextually appropriate.
+-- TODO: More unit testing should be done with a comprehensive calendar table "solved" in another RDBMS that supports datediff natively for comparing against the results of this macro.
+#}
+{% macro sqlite__datediff(first_date, second_date, datepart) -%}
+    {% set datepart = datepart.lower() %}
     {% if datepart == 'year' %}
-        (strftime('%Y', {{second_date}}) - strftime('%Y', {{first_date}}))
-    {#
-    {% elif datepart == 'quarter' %}
-        ({{ datediff(first_date, second_date, 'year') }} * 4 + date_part('quarter', ({{second_date}})::date) - date_part('quarter', ({{first_date}})::date))
-    #}
+        (strftime('%Y', {{ second_date }}) - strftime('%Y', {{ first_date }}))
     {% elif datepart == 'month' %}
-        (({{ datediff(first_date, second_date, 'year') }} * 12 + strftime('%m', {{second_date}})) - strftime('%m', {{first_date}}))
+        ((strftime('%Y', {{ second_date }}) - strftime('%Y', {{ first_date }})) * 12) +
+        (strftime('%m', {{ second_date }}) - strftime('%m', {{ first_date }}))
     {% elif datepart == 'day' %}
-        (floor(cast(strftime('%s', {{second_date}}) - strftime('%s', {{first_date}}) as real) / 86400) +
-        case when {{second_date}} <= strftime('%Y-%m-%d 23:59:59.999999', {{first_date}}) then -1 else 0 end)
+        CASE
+            WHEN 
+                ((strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 86400.0) >= 0
+            THEN CEIL(
+                (strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 86400.0
+            )
+            ELSE FLOOR(
+                (strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 86400.0
+            )
+        END
     {% elif datepart == 'week' %}
-        ({{ datediff(first_date, second_date, 'day') }} / 7 + case
-            when strftime('%w', {{first_date}}) <= strftime('%w', {{second_date}}) then
-                case when {{first_date}} <= {{second_date}} then 0 else -1 end
-            else
-                case when {{first_date}} <= {{second_date}} then 1 else 0 end
-        end)
+        CASE
+            WHEN 
+                ((strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 604800.0) >= 0.285715
```

Comment:
Interestingly enough, `1 / 0.285715 = 3.49999`
Which does seem to reflect 7 days of the week if each day contributes 0.5 to that value. 
So this empirical value of 0.28517 seems to reflect the dividing line between two scenarios: when the two dates being measured for their weekly difference are greater than one day apart from each other, or within one day.

If greater than one day apart, we use CEIL to ensure the week difference is 1
If less than one day apart, we use FLOOR to ensure the week difference is 0.

I don't know whether this is a coincidence or if it mathematically makes sense. 

I feel like I need to run more tests, but making a calendar table at the millisecond grain will need to be done carefully to keep the row count low, yet still satisfy additional testing. Or maybe the tests that are setup already are already covering all edge cases. Only thing I can do is try to run more tests that seem different than what we have already and try to break this macro.

Reply:
If I'm reading it correctly, the implementation for the postgres adapter considers whether the 2nd date "rolls over" into the following week, as determined by whether day of week for second_date is less than that for first_date. So it doesn't have to do with a two-day threshold, it has to do with the fact that 2019-12-31 is a Tuesday, so anything from 2020-01-02 through 2020-01-06 would be considered to have a week difference of 0.

https://github.com/dbt-labs/dbt-postgres/blob/118fbbdc26376dd13cf09b43b31e5eba4235bbe8/dbt/include/postgres/macros/utils/datediff.sql#L11-L17

It's tricky for sqlite because there's no easy way to do a "day of week" calculation, at least with the built-in functions. I wonder if we could find a date library to do this. sqlean, which we're already using, doesn't include any date functions, unfortunately.

Reply:
> So it doesn't have to do with a two-day threshold, it has to do with the fact that 2019-12-31 is a Tuesday, so anything from 2020-01-02 through 2020-01-06 would be considered to have a week difference of 0.

That makes perfect sense now. Thank you for this explanation! 

Also thanks for linking the dbt-postgres implementation, it's good that we reference these existing implementations because I find the edge cases exemplified by the unit tests aren't capturing everything we need yet, but seeing their implementation helps close the gap and shed light.

Regarding sqlite and "day of week", we don't have access to postgres' `dow`, but we can use sqlite's "%w" string format option: https://www.sqlite.org/lang_datefunc.html

```
select cast(strftime('%w', datetime('now')) as integer);
(1,)
```

Using this we could update our `week` part to something like:

```
{% elif datepart == 'week' %}
        -- Calculate the initial week difference by finding the difference in days and dividing by 7
        (
            ((strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 86400) / 7 +
            -- Adjust based on the days of the week to ensure correct week boundaries
            CASE
                WHEN strftime('%w', {{ first_date }}) <= strftime('%w', {{ second_date }})
                THEN CASE
                        WHEN strftime('%s', {{ first_date }}) <= strftime('%s', {{ second_date }})
                        THEN 0
                        ELSE -1
                    END
                ELSE CASE
                        WHEN strftime('%s', {{ first_date }}) <= strftime('%s', {{ second_date }})
                        THEN 1
                        ELSE 0
                    END
            END
        )
```

Though for other parts of the macro, I am now questioning the use of CEIL v FLOOR knowing it likely could go wrong somewhere. I'll look into the postgres native implementation for datediff and see if I can learn a bit more from that to refine this PR.

Reply:
I also think I should consider using the code that was already adapted from postgresql to the left because it might be more performant in some areas, and also reads more like the postgresql one which would be a common reference point. One example of this is the case statement only being applied on the second operand used for adjustment, not on the whole expression, which avoids calling the same functions (cast and floor/ceil) multiple times on the same expression which is wasteful.
 

For example,